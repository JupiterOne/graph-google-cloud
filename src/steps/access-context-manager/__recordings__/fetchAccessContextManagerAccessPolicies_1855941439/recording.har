{
  "log": {
    "_recordingName": "fetchAccessContextManagerAccessPolicies",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 739,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "739"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 300,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 572,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 572,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Wed, 02 Jun 2021 14:22:23 GMT"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 511,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-02T14:22:23.123Z",
        "time": 212,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 212
        }
      },
      {
        "_id": "e7ae7a9d3285d95d4330a7c239aa61b7",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/5.0.2 gl-node/14.15.4 auth/7.1.0"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/5.0.2 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "accesscontextmanager.googleapis.com"
            }
          ],
          "headersSize": 581,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "parent",
              "value": "organizations/958457776463"
            }
          ],
          "url": "https://accesscontextmanager.googleapis.com/v1/accessPolicies?parent=organizations%2F958457776463"
        },
        "response": {
          "bodySize": 369,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 369,
            "text": "{\"accessPolicies\":[{\"name\":\"accessPolicies/368605946207\",\"parent\":\"organizations/958457776463\",\"title\":\"default policy\",\"etag\":\"64186ca0c82b01b4\"}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Wed, 02 Jun 2021 14:22:24 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 488,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-02T14:22:23.374Z",
        "time": 829,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 829
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
