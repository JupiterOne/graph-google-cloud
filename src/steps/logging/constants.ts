export const STEP_LOGGING_PROJECT_SINKS = 'fetch-logging-project-sinks';
export const STEP_CREATE_LOGGING_PROJECT_SINK_BUCKET_RELATIONSHIPS =
  'build-logging-project-sink-bucket-relationships';
export const STEP_LOGGING_METRICS = 'fetch-logging-metrics';

export const LOGGING_PROJECT_SINK_ENTITY_TYPE = 'google_logging_project_sink';
export const LOGGING_PROJECT_SINK_ENTITY_CLASS = 'Logs';

export const LOGGING_METRIC_ENTITY_TYPE = 'google_logging_metric';
export const LOGGING_METRIC_ENTITY_CLASS = 'Configuration';

export const RELATIONSHIP_TYPE_PROJECT_SINK_USES_STORAGE_BUCKET =
  'google_logging_project_sink_uses_storage_bucket';

export const RELATIONSHIP_TYPE_METRIC_HAS_ALERT_POLICY =
  'google_logging_metric_has_monitoring_alert_policy';

export const IngestionSources = {
  LOGGING_PROJECT_SINKS: 'logging-project-sinks',
  LOGGING_METRICS: 'logging-metrics',
};

export const LoggingIngestionConfig = {
  [IngestionSources.LOGGING_PROJECT_SINKS]: {
    title: 'Google Logging Project Sinks',
    description: 'Destinations for log entries export.',
    defaultsToDisabled: true,
  },
  [IngestionSources.LOGGING_METRICS]: {
    title: 'Google Logging Metrics',
    description: 'Custom metrics from log data.',
    defaultsToDisabled: true,
  },
};
