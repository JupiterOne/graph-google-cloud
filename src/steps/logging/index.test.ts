import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import { fetchSinks, fetchMetrics } from '.';
import { fetchAlertPolicies } from '../monitoring';
import {
  CLOUD_STORAGE_BUCKET_ENTITY_TYPE,
  fetchStorageBuckets,
} from '../storage';
import { integrationConfig } from '../../../test/config';
import { Recording, setupGoogleCloudRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../types';

import {
  LOGGING_METRIC_ENTITY_TYPE,
  LOGGING_PROJECT_SINK_ENTITY_TYPE,
  RELATIONSHIP_TYPE_METRIC_HAS_ALERT_POLICY,
  RELATIONSHIP_TYPE_PROJECT_SINK_USES_STORAGE_BUCKET,
} from './constants';
import { MONITORING_ALERT_POLICY_TYPE } from '../monitoring/constants';

describe('#fetchProjectSinks', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchProjectSinks',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchStorageBuckets(context);
    await fetchSinks(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === CLOUD_STORAGE_BUCKET_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['DataStore'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_storage_bucket' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          storageClass: { type: 'string' },
          encrypted: { const: true },
          encryptionKeyRef: { type: 'string' },
          kmsKeyName: { type: 'string' },
          uniformBucketLevelAccess: { type: 'boolean' },
          retentionPolicyEnabled: { type: 'boolean' },
          retentionPeriod: { type: 'string' },
          retentionDate: { type: 'string' },
          public: { type: 'boolean' },
          accessLevel: { type: 'string' },
          accessLevelIsConditional: { type: 'boolean' },
          accessLevelCondition: { type: 'string' },
          classification: { const: null },
          etag: { type: 'string' },
          webLink: { type: 'string' },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === LOGGING_PROJECT_SINK_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Logs'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_logging_project_sink' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          destination: { type: 'string' },
          filter: { type: 'string' },
          webLink: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_PROJECT_SINK_USES_STORAGE_BUCKET,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'USES' },
          _type: { const: 'google_logging_project_sink_uses_storage_bucket' },
        },
      },
    });
  });
});

describe('#fetchMetrics', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchMetrics',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAlertPolicies(context);
    await fetchMetrics(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === MONITORING_ALERT_POLICY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Policy'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_monitoring_alert_policy' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          title: { type: 'string' },
          summary: { type: 'string' },
          content: { type: 'string' },
          conditionFilters: {
            type: 'array',
            items: { type: 'string' },
          },
          enabled: { type: 'boolean' },
          webLink: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === LOGGING_METRIC_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Configuration'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_logging_metric' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          filter: { type: 'string' },
          webLink: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_METRIC_HAS_ALERT_POLICY,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: { const: 'google_logging_metric_has_monitoring_alert_policy' },
        },
      },
    });
  });
});
