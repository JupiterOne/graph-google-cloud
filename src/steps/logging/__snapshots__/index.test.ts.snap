// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchMetrics should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Policy",
      ],
      "_key": "projects/j1-gc-integration-dev-v2/alertPolicies/8900061597970753716",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "combiner": "OR",
            "conditions": Array [
              Object {
                "conditionThreshold": Object {
                  "aggregations": Array [
                    Object {
                      "alignmentPeriod": "300s",
                      "perSeriesAligner": "ALIGN_RATE",
                    },
                  ],
                  "comparison": "COMPARISON_GT",
                  "duration": "0s",
                  "filter": "metric.type=\\"logging.googleapis.com/user/my-example-metric\\" AND resource.type=\\"metric\\"",
                  "trigger": Object {
                    "count": 1,
                  },
                },
                "displayName": "test condition",
                "name": "projects/j1-gc-integration-dev-v2/alertPolicies/8900061597970753716/conditions/8900061597970755317",
              },
            ],
            "creationRecord": Object {
              "mutateTime": "2021-02-19T13:41:42.161511391Z",
              "mutatedBy": "j1-gc-integration-dev-v2-tf@j1-gc-integration-dev-v2.iam.gserviceaccount.com",
            },
            "displayName": "Example Alert Policy",
            "enabled": true,
            "mutationRecord": Object {
              "mutateTime": "2021-02-19T13:41:42.161511391Z",
              "mutatedBy": "j1-gc-integration-dev-v2-tf@j1-gc-integration-dev-v2.iam.gserviceaccount.com",
            },
            "name": "projects/j1-gc-integration-dev-v2/alertPolicies/8900061597970753716",
          },
        },
      ],
      "_type": "google_monitoring_alert_policy",
      "conditionFilters": Array [
        "metric.type=\\"logging.googleapis.com/user/my-example-metric\\" AND resource.type=\\"metric\\"",
      ],
      "content": "",
      "createdOn": 1613742102161,
      "displayName": "Example Alert Policy",
      "enabled": true,
      "name": "Example Alert Policy",
      "summary": "Alert policy that is triggered based on some metric",
      "title": "Alert policy",
      "updatedOn": 1613742102161,
      "webLink": "https://console.cloud.google.com/monitoring/alerting/policies/8900061597970753716?project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "Configuration",
      ],
      "_key": "my-example-metric",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-02-17T23:05:06.358873155Z",
            "filter": "protoPayload.methodName=\\"SetIamPolicy\\" AND protoPayload.serviceData.policyDelta.auditConfigDeltas:*",
            "labelExtractors": Object {
              "occurences": "EXTRACT(textPayload)",
            },
            "metricDescriptor": Object {
              "displayName": "My metric",
              "labels": Array [
                Object {
                  "key": "occurences",
                  "valueType": "INT64",
                },
              ],
              "metricKind": "DELTA",
              "name": "projects/j1-gc-integration-dev-v2/metricDescriptors/logging.googleapis.com/user/my-example-metric",
              "type": "logging.googleapis.com/user/my-example-metric",
              "unit": "1",
              "valueType": "INT64",
            },
            "name": "my-example-metric",
            "updateTime": "2021-02-17T23:05:06.358873155Z",
          },
        },
      ],
      "_type": "google_logging_metric",
      "createdOn": 1613603106358,
      "displayName": "my-example-metric",
      "filter": "protoPayload.methodName=\\"SetIamPolicy\\" AND protoPayload.serviceData.policyDelta.auditConfigDeltas:*",
      "name": "my-example-metric",
      "updatedOn": 1613603106358,
      "webLink": "https://console.cloud.google.com/logs/metrics?project=j1-gc-integration-dev-v2",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "my-example-metric",
      "_key": "my-example-metric|has|projects/j1-gc-integration-dev-v2/alertPolicies/8900061597970753716",
      "_toEntityKey": "projects/j1-gc-integration-dev-v2/alertPolicies/8900061597970753716",
      "_type": "google_logging_metric_has_monitoring_alert_policy",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "google_monitoring_alert_policy",
    "google_logging_metric",
    "google_logging_metric_has_monitoring_alert_policy",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;

exports[`#fetchProjectSinks should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "DataStore",
      ],
      "_key": "bucket:gcf-sources-711888229551-us-central1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "cors": Array [
              Object {
                "method": Array [
                  "GET",
                ],
                "origin": Array [
                  "https://*.cloud.google.com",
                  "https://*.corp.google.com",
                  "https://*.corp.google.com:*",
                ],
              },
            ],
            "etag": "CAE=",
            "iamConfiguration": Object {
              "bucketPolicyOnly": Object {
                "enabled": true,
                "lockedTime": "2021-05-19T00:04:30.131Z",
              },
              "uniformBucketLevelAccess": Object {
                "enabled": true,
                "lockedTime": "2021-05-19T00:04:30.131Z",
              },
            },
            "id": "gcf-sources-711888229551-us-central1",
            "kind": "storage#bucket",
            "location": "US-CENTRAL1",
            "locationType": "region",
            "metageneration": "1",
            "name": "gcf-sources-711888229551-us-central1",
            "projectNumber": "711888229551",
            "selfLink": "https://www.googleapis.com/storage/v1/b/gcf-sources-711888229551-us-central1",
            "storageClass": "STANDARD",
            "timeCreated": "2021-02-18T00:04:30.131Z",
            "updated": "2021-02-18T00:04:30.131Z",
          },
        },
      ],
      "_type": "google_storage_bucket",
      "accessLevel": "private",
      "accessLevelCondition": undefined,
      "accessLevelIsConditional": false,
      "classification": null,
      "createdOn": 1613606670131,
      "displayName": "gcf-sources-711888229551-us-central1",
      "encrypted": true,
      "encryptionKeyRef": undefined,
      "etag": "CAE=",
      "id": "gcf-sources-711888229551-us-central1",
      "kmsKeyName": undefined,
      "location": "US-CENTRAL1",
      "name": "gcf-sources-711888229551-us-central1",
      "public": false,
      "retentionDate": undefined,
      "retentionPeriod": undefined,
      "retentionPolicyEnabled": undefined,
      "storageClass": "STANDARD",
      "uniformBucketLevelAccess": true,
      "updatedOn": 1613606670131,
      "webLink": "https://console.cloud.google.com/storage/browser/gcf-sources-711888229551-us-central1;tab=objects?forceOnBucketsSortingFiltering=false&project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "DataStore",
      ],
      "_key": "bucket:j1-gc-integration-dev-v2-sink-logging-bucket",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "etag": "CAE=",
            "iamConfiguration": Object {
              "bucketPolicyOnly": Object {
                "enabled": false,
              },
              "uniformBucketLevelAccess": Object {
                "enabled": false,
              },
            },
            "id": "j1-gc-integration-dev-v2-sink-logging-bucket",
            "kind": "storage#bucket",
            "location": "US",
            "locationType": "multi-region",
            "metageneration": "1",
            "name": "j1-gc-integration-dev-v2-sink-logging-bucket",
            "projectNumber": "711888229551",
            "selfLink": "https://www.googleapis.com/storage/v1/b/j1-gc-integration-dev-v2-sink-logging-bucket",
            "storageClass": "STANDARD",
            "timeCreated": "2021-02-18T00:31:30.869Z",
            "updated": "2021-02-18T00:31:30.869Z",
          },
        },
      ],
      "_type": "google_storage_bucket",
      "accessLevel": "SubjectToObjectACLs",
      "accessLevelCondition": "Subject to object ACLs",
      "accessLevelIsConditional": true,
      "classification": null,
      "createdOn": 1613608290869,
      "displayName": "j1-gc-integration-dev-v2-sink-logging-bucket",
      "encrypted": true,
      "encryptionKeyRef": undefined,
      "etag": "CAE=",
      "id": "j1-gc-integration-dev-v2-sink-logging-bucket",
      "kmsKeyName": undefined,
      "location": "US",
      "name": "j1-gc-integration-dev-v2-sink-logging-bucket",
      "public": false,
      "retentionDate": undefined,
      "retentionPeriod": undefined,
      "retentionPolicyEnabled": undefined,
      "storageClass": "STANDARD",
      "uniformBucketLevelAccess": false,
      "updatedOn": 1613608290869,
      "webLink": "https://console.cloud.google.com/storage/browser/j1-gc-integration-dev-v2-sink-logging-bucket;tab=objects?forceOnBucketsSortingFiltering=false&project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "DataStore",
      ],
      "_key": "bucket:j1-gc-integration-dev-v2cloudfunctions",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "etag": "CAE=",
            "iamConfiguration": Object {
              "bucketPolicyOnly": Object {
                "enabled": false,
              },
              "uniformBucketLevelAccess": Object {
                "enabled": false,
              },
            },
            "id": "j1-gc-integration-dev-v2cloudfunctions",
            "kind": "storage#bucket",
            "location": "US",
            "locationType": "multi-region",
            "metageneration": "1",
            "name": "j1-gc-integration-dev-v2cloudfunctions",
            "projectNumber": "711888229551",
            "selfLink": "https://www.googleapis.com/storage/v1/b/j1-gc-integration-dev-v2cloudfunctions",
            "storageClass": "STANDARD",
            "timeCreated": "2021-02-17T23:05:05.322Z",
            "updated": "2021-02-17T23:05:05.322Z",
          },
        },
      ],
      "_type": "google_storage_bucket",
      "accessLevel": "SubjectToObjectACLs",
      "accessLevelCondition": "Subject to object ACLs",
      "accessLevelIsConditional": true,
      "classification": null,
      "createdOn": 1613603105322,
      "displayName": "j1-gc-integration-dev-v2cloudfunctions",
      "encrypted": true,
      "encryptionKeyRef": undefined,
      "etag": "CAE=",
      "id": "j1-gc-integration-dev-v2cloudfunctions",
      "kmsKeyName": undefined,
      "location": "US",
      "name": "j1-gc-integration-dev-v2cloudfunctions",
      "public": false,
      "retentionDate": undefined,
      "retentionPeriod": undefined,
      "retentionPolicyEnabled": undefined,
      "storageClass": "STANDARD",
      "uniformBucketLevelAccess": false,
      "updatedOn": 1613603105322,
      "webLink": "https://console.cloud.google.com/storage/browser/j1-gc-integration-dev-v2cloudfunctions;tab=objects?forceOnBucketsSortingFiltering=false&project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "DataStore",
      ],
      "_key": "bucket:us.artifacts.j1-gc-integration-dev-v2.appspot.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "etag": "CAE=",
            "iamConfiguration": Object {
              "bucketPolicyOnly": Object {
                "enabled": false,
              },
              "uniformBucketLevelAccess": Object {
                "enabled": false,
              },
            },
            "id": "us.artifacts.j1-gc-integration-dev-v2.appspot.com",
            "kind": "storage#bucket",
            "location": "US",
            "locationType": "multi-region",
            "metageneration": "1",
            "name": "us.artifacts.j1-gc-integration-dev-v2.appspot.com",
            "projectNumber": "711888229551",
            "selfLink": "https://www.googleapis.com/storage/v1/b/us.artifacts.j1-gc-integration-dev-v2.appspot.com",
            "storageClass": "STANDARD",
            "timeCreated": "2021-02-18T00:25:22.127Z",
            "updated": "2021-02-18T00:25:22.127Z",
          },
        },
      ],
      "_type": "google_storage_bucket",
      "accessLevel": "SubjectToObjectACLs",
      "accessLevelCondition": "Subject to object ACLs",
      "accessLevelIsConditional": true,
      "classification": null,
      "createdOn": 1613607922127,
      "displayName": "us.artifacts.j1-gc-integration-dev-v2.appspot.com",
      "encrypted": true,
      "encryptionKeyRef": undefined,
      "etag": "CAE=",
      "id": "us.artifacts.j1-gc-integration-dev-v2.appspot.com",
      "kmsKeyName": undefined,
      "location": "US",
      "name": "us.artifacts.j1-gc-integration-dev-v2.appspot.com",
      "public": false,
      "retentionDate": undefined,
      "retentionPeriod": undefined,
      "retentionPolicyEnabled": undefined,
      "storageClass": "STANDARD",
      "uniformBucketLevelAccess": false,
      "updatedOn": 1613607922127,
      "webLink": "https://console.cloud.google.com/storage/browser/us.artifacts.j1-gc-integration-dev-v2.appspot.com;tab=objects?forceOnBucketsSortingFiltering=false&project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "Logs",
      ],
      "_key": "google_logging_log_sink:my-instance-sink",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-02-18T00:31:31.329464207Z",
            "description": "an example sink",
            "destination": "storage.googleapis.com/j1-gc-integration-dev-v2-sink-logging-bucket",
            "name": "my-instance-sink",
            "updateTime": "2021-02-18T00:31:31.329464207Z",
            "writerIdentity": "serviceAccount:cloud-logs@system.gserviceaccount.com",
          },
        },
      ],
      "_type": "google_logging_project_sink",
      "createdOn": 1613608291329,
      "description": "an example sink",
      "destination": "storage.googleapis.com/j1-gc-integration-dev-v2-sink-logging-bucket",
      "displayName": "my-instance-sink",
      "filter": undefined,
      "name": "my-instance-sink",
      "updatedOn": 1613608291329,
      "webLink": "https://console.cloud.google.com/logs/router?project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "Logs",
      ],
      "_key": "google_logging_log_sink:_Required",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "destination": "logging.googleapis.com/projects/j1-gc-integration-dev-v2/locations/global/buckets/_Required",
            "filter": "LOG_ID(\\"cloudaudit.googleapis.com/activity\\") OR LOG_ID(\\"externalaudit.googleapis.com/activity\\") OR LOG_ID(\\"cloudaudit.googleapis.com/system_event\\") OR LOG_ID(\\"externalaudit.googleapis.com/system_event\\") OR LOG_ID(\\"cloudaudit.googleapis.com/access_transparency\\") OR LOG_ID(\\"externalaudit.googleapis.com/access_transparency\\")",
            "name": "_Required",
          },
        },
      ],
      "_type": "google_logging_project_sink",
      "createdOn": undefined,
      "destination": "logging.googleapis.com/projects/j1-gc-integration-dev-v2/locations/global/buckets/_Required",
      "displayName": "_Required",
      "filter": "LOG_ID(\\"cloudaudit.googleapis.com/activity\\") OR LOG_ID(\\"externalaudit.googleapis.com/activity\\") OR LOG_ID(\\"cloudaudit.googleapis.com/system_event\\") OR LOG_ID(\\"externalaudit.googleapis.com/system_event\\") OR LOG_ID(\\"cloudaudit.googleapis.com/access_transparency\\") OR LOG_ID(\\"externalaudit.googleapis.com/access_transparency\\")",
      "name": "_Required",
      "updatedOn": undefined,
      "webLink": "https://console.cloud.google.com/logs/router?project=j1-gc-integration-dev-v2",
    },
    Object {
      "_class": Array [
        "Logs",
      ],
      "_key": "google_logging_log_sink:_Default",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "destination": "logging.googleapis.com/projects/j1-gc-integration-dev-v2/locations/global/buckets/_Default",
            "filter": "NOT LOG_ID(\\"cloudaudit.googleapis.com/activity\\") AND NOT LOG_ID(\\"externalaudit.googleapis.com/activity\\") AND NOT LOG_ID(\\"cloudaudit.googleapis.com/system_event\\") AND NOT LOG_ID(\\"externalaudit.googleapis.com/system_event\\") AND NOT LOG_ID(\\"cloudaudit.googleapis.com/access_transparency\\") AND NOT LOG_ID(\\"externalaudit.googleapis.com/access_transparency\\")",
            "name": "_Default",
          },
        },
      ],
      "_type": "google_logging_project_sink",
      "createdOn": undefined,
      "destination": "logging.googleapis.com/projects/j1-gc-integration-dev-v2/locations/global/buckets/_Default",
      "displayName": "_Default",
      "filter": "NOT LOG_ID(\\"cloudaudit.googleapis.com/activity\\") AND NOT LOG_ID(\\"externalaudit.googleapis.com/activity\\") AND NOT LOG_ID(\\"cloudaudit.googleapis.com/system_event\\") AND NOT LOG_ID(\\"externalaudit.googleapis.com/system_event\\") AND NOT LOG_ID(\\"cloudaudit.googleapis.com/access_transparency\\") AND NOT LOG_ID(\\"externalaudit.googleapis.com/access_transparency\\")",
      "name": "_Default",
      "updatedOn": undefined,
      "webLink": "https://console.cloud.google.com/logs/router?project=j1-gc-integration-dev-v2",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "USES",
      "_fromEntityKey": "google_logging_log_sink:my-instance-sink",
      "_key": "google_logging_log_sink:my-instance-sink|uses|bucket:j1-gc-integration-dev-v2-sink-logging-bucket",
      "_toEntityKey": "bucket:j1-gc-integration-dev-v2-sink-logging-bucket",
      "_type": "google_logging_project_sink_uses_storage_bucket",
      "displayName": "USES",
    },
  ],
  "encounteredTypes": Array [
    "google_storage_bucket",
    "google_logging_project_sink",
    "google_logging_project_sink_uses_storage_bucket",
  ],
  "numCollectedEntities": 7,
  "numCollectedRelationships": 1,
}
`;
